### PR TITLE
Stabilizing some test functions with set random seeds

### PR DESCRIPTION
**Pull request by pjsjongsung**
This pull request was created according to #2928 
I have added a decorator to fix numpy's random seed.
To use the decorator, I have moved the global variable in `dipy/segment/tests/test_mrf.py` to a function that creates it.

Are there any more changes I should make/revert? Also, do we want to add this decorator to every test function that has a numpy random function?

Thanks!

### CODE REVIEW

**Comment by skoudoro** [comment]
Hi @pjsjongsung,

Thank you for taking care of this! I like the decorator approach. For sure there are other functions like the ones in tracking module but for now, it is ok. let's just focus on ICM.

File: dipy/testing/decorators.py [hunk 1]
```diff
@@ -73,3 +73,18 @@ def test_with_xvfb(*args, **kwargs):
...
+
+def set_random_seed(func):
+    """Decorator to set equal seeds.
+
+    This will help us avoid random errors
+    that occur with low probabilitiy during testing.
+
+    """
+    def test_with_seed(*args, **kwargs):
+        np.random.seed(0)
```
Comment (skoudoro) [inline dipy/testing/decorators.py:86]: instead of zero, I would recommend to put a number., whatever it is.... something like this year (`np.random.seed(20223)`)

File: dipy/testing/decorators.py [hunk 2]
```diff
@@ -73,3 +73,18 @@ def test_with_xvfb(*args, **kwargs):
     # Plant it back in and return the new function:
     test_with_xvfb.__name__ = fname
     return test_with_xvfb if not is_windows else my_test
+
+
+def set_random_seed(func):
+    """Decorator to set equal seeds.
+
+    This will help us avoid random errors
+    that occur with low probabilitiy during testing.
```
Comment (jhlegarreta) [inline dipy/testing/decorators.py:82]: The primary purpose of this is not to prevent errors, but to make results reproducible.

File: dipy/testing/decorators.py [hunk 3]
```diff
@@ -73,3 +73,18 @@ def test_with_xvfb(*args, **kwargs):
     # Plant it back in and return the new function:
     test_with_xvfb.__name__ = fname
     return test_with_xvfb if not is_windows else my_test
+
+
+def set_random_seed(func):
+    """Decorator to set equal seeds.
```
Comment (jhlegarreta) [inline dipy/testing/decorators.py:79]: Decorator to use a fixed value for the random generator seed.

File: dipy/testing/decorators.py [hunk 4]
```diff
@@ -73,3 +73,18 @@ def test_with_xvfb(*args, **kwargs):
...
+
+    This will help us avoid random errors
+    that occur with low probabilitiy during testing.
+
+    """
+    def test_with_seed(*args, **kwargs):
+        np.random.seed(2023)
```
Comment (jhlegarreta) [inline dipy/testing/decorators.py:86]: This has been deprecated in favor of:
```
rng = np.random.default_rng(<seed_value>)
rng.random(<value_count>)
```

so the decorator should be changed. Not sure how hard it is to make the `rng` variable be exposed to the testing method.

https://numpy.org/doc/stable/reference/random/generator.html
https://numpy.org/neps/nep-0019-rng-policy.html

File: dipy/testing/decorators.py [hunk 5]
```diff
@@ -73,3 +73,18 @@ def test_with_xvfb(*args, **kwargs):
     # Plant it back in and return the new function:
     test_with_xvfb.__name__ = fname
     return test_with_xvfb if not is_windows else my_test
+
+
+def set_random_seed(func):
```
Comment (jhlegarreta) [inline dipy/testing/decorators.py:78]: I would rename this to something like `fix_random_number_generator`, as the seed is fixed, and cannot be fixed from the outside. The name should then be made extensive to the inner method.

File: dipy/segment/tests/test_mrf.py [hunk 6]
```diff
@@ -28,32 +29,42 @@
...
-square_1[42:213, 42:213, :] = temp_1
-temp_2 = np.random.randint(1, 21, size=(114, 114, 3))
-temp_2 = np.where(temp_2 < 19, 2, 1)
-square_1[71:185, 71:185, :] = temp_2
-temp_3 = np.random.randint(1, 21, size=(58, 58, 3))
-temp_3 = np.where(temp_3 < 20, 3, 1)
-square_1[99:157, 99:157, :] = temp_3
+
+@set_random_seed
+def create_square_uniform():
```
Comment (jhlegarreta) [inline dipy/segment/tests/test_mrf.py:34]: Why is this change part of the commit?

**Review by jhlegarreta — CHANGES_REQUESTED** [review]
Thanks for giving this a try @pjsjongsung. A few comments inline.

File: dipy/segment/tests/test_mrf.py (hunk 6 above)
Comment (pjsjongsung) [reply dipy/segment/tests/test_mrf.py:34]: Because I had to add a decorator when creating the global variables. Adding the decorator to the test functions will be too late. Before, it was just defining them globally, which I do not think I can add a decorator to it.

Comment (jhlegarreta) [reply dipy/segment/tests/test_mrf.py:34]: OK, then please explain in the commit message. Thanks Jong Sung.

File: dipy/testing/decorators.py (hunk 4 above)
Comment (pjsjongsung) [reply dipy/testing/decorators.py:86]: Yeah I understand it is deprecated, but as you said I am not sure how to use the decorator with the rng variable. Alternatively, we could just change each random function individually.

Comment (jhlegarreta) [reply dipy/testing/decorators.py:86]: The risk of using a deprecated value is the removal in a future `NumPy` version, so better IMO it is better to stick to the current convention now.

Comment (skoudoro) [reply dipy/testing/decorators.py:86]: I did not know that `np.random.seed` has been deprecated. Thank you for the info @jhlegarreta .

We need to update that in the whole code based. I prefer the new standard.

**Comment by pjsjongsung** [comment]
I have changed the icm tests to the random generator. If this looks ok, we could change all random functions accordingly?

File: dipy/segment/tests/test_mrf.py [hunk 7]
```diff
@@ -28,32 +29,47 @@
...
-temp_1 = np.where(temp_1 < 20, 1, 3)
-square_1[42:213, 42:213, :] = temp_1
-temp_2 = np.random.randint(1, 21, size=(114, 114, 3))
-temp_2 = np.where(temp_2 < 19, 2, 1)
-square_1[71:185, 71:185, :] = temp_2
-temp_3 = np.random.randint(1, 21, size=(58, 58, 3))
-temp_3 = np.where(temp_3 < 20, 3, 1)
-square_1[99:157, 99:157, :] = temp_3
+# Setting seed value
+seed_v = 1234
```
Comment (jhlegarreta) [inline dipy/segment/tests/test_mrf.py:33]: Is this needed?

Comment (pjsjongsung) [reply dipy/segment/tests/test_mrf.py]: Meant to remove that. Thank you!

**Comment by pjsjongsung** [comment]
It seems to be disrupting assertions of various tests using np.random.seed. I will try changing all of them to generators, and if that doesn't work, we should not change this until we figure out what the reason for the assertion errors are.

File: dipy/testing/decorators.py [hunk 8]
```diff
@@ -73,3 +73,18 @@ def test_with_xvfb(*args, **kwargs):
...
     return test_with_xvfb if not is_windows else my_test
+
+
+def fix_random_number_generator(func):
+    """Decorator to use a fixed value for the random generator seed.
+
+    This will make the tests that use random functions reproducible.
+
+    """
+    def wrapper_with_fixed_seed(*args, **kwargs):
```
Comment (jhlegarreta) [inline dipy/testing/decorators.py:84]: I'd change this name so that it more closely follows the callable one.

**Review by jhlegarreta — COMMENTED** [review]
Thanks for the effort: I would split this into two/three commits so that it becomes crystal clear the purpose, benefit, effects of each commit (otherwise, it becomes hard to find and explain things when the changes are revisited in the future):
- The first one would transition to the current NumPy recommended way to use the random generator across the whole code base.
- The decorator is added and used across all tests (could be done together with the first commit)
- A last commit would try to address the ICM test failure/instability.

**Comment by pjsjongsung** [comment]
I am working on the new first commit, which is transitioning everything to a numpy recommended way of random generator. However, there seems to be tests that were only working on particular seeds, and is now disrupted by the different random values. (Even with the same seed, np.random.seed and random generator have different results). I will keep looking at what the reasons might be, but it seems it is not an easy solution.

For now, I will create a separate PR for test stability of ICM.

**Comment by jhlegarreta** [comment]
Re https://github.com/dipy/dipy/pull/2929#issuecomment-1768504561

> (Even with the same seed, np.random.seed and random generator have different results)

Maybe this is expected/documented somewhere or can be asked to the NumPy folks.

Thanks for effort @pjsjongsung.

**Comment by pjsjongsung** [comment]
> Re [#2929 (comment)](https://github.com/dipy/dipy/pull/2929#issuecomment-1768504561)
> 
> > (Even with the same seed, np.random.seed and random generator have different results)
> 
> Maybe this is expected/documented somewhere or can be asked to the NumPy folks.
It seems they have changed Bit Generator that is used to create random values. So until they change that again, (which I believe  is far away since they changed it only once ever) the seed values should be stable enough.

**Comment by pjsjongsung** [comment]
It seems all the tests related to the files I changed are passing.

**Review by jhlegarreta — CHANGES_REQUESTED** [review]
> It seems they have changed Bit Generator that is used to create random values. So until they change that again, (which I believe is far away since they changed it only once ever) the seed values should be stable enough.

OK. Thanks for the investigation Jong Sung.

> It seems all the tests related to the files I changed are passing.

:ok_hand: Just curious: then, why were you reporting that some tests were failing in https://github.com/dipy/dipy/pull/2929#issuecomment-1753653282? Maybe because the seed was set to a different value wrt the previous one?

> fix_random_number_generator

I now realize that it is not the best name (my bad, as I proposed it): can it be changed to `set_random_number_generator`, the private method just be `_set_random_number_generator`, and its descendant be `_set_random_number_generator_wrapper`?

Please, use a single commit: `TEST: Changing numpy random usage in tests to random generator` with the meaningful commit body message you wrote, the rest of the commit are intermediate work and are not informative when navigating the git history. You will need to squash all commits, and push force.

**Comment by pjsjongsung** [comment]
> 👌 Just curious: then, why were you reporting that some tests were failing in [#2929 (comment)](https://github.com/dipy/dipy/pull/2929#issuecomment-1753653282)? Maybe because the seed was set to a different value wrt the previous one?

So `numpy.random.seed` seems to use a different random number generator than `numpy.random.Generator`. Thus even with the same seed, completely different values were being created. So there needed to be some searching on new seed values that worked.
 
> > fix_random_number_generator
> 
> I now realize that it is not the best name (my bad, as I proposed it): can it be changed to `set_random_number_generator`, the private method just be `_set_random_number_generator`, and its descendant be `_set_random_number_generator_wrapper`?
> 

> Please, use a single commit: `TEST: Changing numpy random usage in tests to random generator` with the meaningful commit body message you wrote, the rest of the commit are intermediate work and are not informative when navigating the git history. You will need to squash all commits, and push force.

Will do.

**Comment by jhlegarreta** [comment]
> So numpy.random.seed seems to use a different random number generator than numpy.random.Generator. Thus even with the same seed, completely different values were being created. So there needed to be some searching on new seed values that worked.

In a quick inspection I saw the same seeds being used/no baselines updated. So if you had to look for new seeds, among the infinite values, how did you find the right values in a finite amount of time :thinking:?

**Comment by pjsjongsung** [comment]
> > So numpy.random.seed seems to use a different random number generator than numpy.random.Generator. Thus even with the same seed, completely different values were being created. So there needed to be some searching on new seed values that worked.
> 
> In a quick inspection I saw the same seeds being used/no baselines updated. So if you had to look for new seeds, among the infinite values, how did you find the right values in a finite amount of time 🤔?

Most of the tests were stable enough to work without change. Others, I had to randomly search a few times, but it worked out. I just ran a for loop with different seed values until it worked :)

**Comment by pjsjongsung** [comment]
Commit squashed and there seems to be no errors from this PR!

**Comment by jhlegarreta** [comment]
:ok_hand: Beautiful @pjsjongsung. Some failing tests seem to be related to the use of random generators, although they were not modified in this patch set:
https://github.com/dipy/dipy/actions/runs/6594238121/job/17917878390?pr=2929#step:9:4397

Do you have an intuition about these?

**Comment by pjsjongsung** [comment]
@jhlegarreta It must be connected to the change in using seeds. The change in seeds are global, so even if the function was not specifically using the seed within the function, it would have been benefiting from a fixed number. (Is my guess)

I have changed the tests that failed and while doing that have changed some other functions that were using some kind of numpy random function but was not using a fixed seed generator.

However, once this PR is merged, I will create another PR to change every numpy random functions used in tests (even if it was not using a seed before) to use some kind of seed.

**Comment by pjsjongsung** [comment]
I think the PR is good to go? The mlp test function is failing for some reason, which does not even use any type of random function. My guess is the change in seeds changed something in the order of the training data being loaded. (IMO it is strange to have a test function on the training accuracy of a DL model after 1 epoch)

**Comment by jhlegarreta** [comment]
> I think the PR is good to go? The mlp test function is failing for some reason, which does not even use any type of random function. My guess is the change in seeds changed something in the order of the training data being loaded.

Have seen the MLP failure; did not spot where the randomness is involved, other than the possible weights, probably internally/out of the sight of the end user. If the seeds were/are being set somehow globally, this may affect the test at issue.

>  (IMO it is strange to have a test function on the training accuracy of a DL model after 1 epoch)

Agreed.

Thoughts:
- Change the test to test only that the model can be instantiated (in a separate commit and/or PR).
- Relax the accuracy tolerance.
- Training an MLP that is not large with simple data should not be that long, so increasing the number of epochs may be reasonable to find a better accuracy.

**Comment by pjsjongsung** [comment]
I have increased the number of epochs to 5 as @jhlegarreta suggested, so it should be more stable than before.

**Comment by jhlegarreta** [comment]
9afe970 Commit message subject prefix, please.

**Comment by pjsjongsung** [comment]
Sorry, my bad. Still getting used to it. Corrected

**Review by jhlegarreta — APPROVED** [review]
Build failure is unrelated.

Thanks for this @pjsjongsung.

**Review by skoudoro — APPROVED** [review]
Thank you for this huge work @pjsjongsung and Thank you for reviewing @jhlegarreta .

Looks good to me too. merging
